subsys/bluetooth/mesh/access.c

```diff
@@ -834,20 +834,24 @@ void bt_mesh_model_tree_walk(struct bt_mesh_model *root,
 {
 	struct bt_mesh_model *m = root;
 	uint32_t depth = 0;
+	bool subtree_visited = false;
 
 	do {
-		if (cb(m, depth, user_data) == BT_MESH_WALK_STOP) {
+		if (!subtree_visited && cb(m, depth, user_data) == BT_MESH_WALK_STOP) {
 			return;
 		}
 #ifdef CONFIG_BT_MESH_MODEL_EXTENSIONS
-		if (m->extends) {
+		if (!subtree_visited && m->extends) {
 			m = m->extends;
 			depth++;
+			subtree_visited = false;
 		} else if (m->flags & BT_MESH_MOD_NEXT_IS_PARENT) {
-			m = m->next->next;
+			m = m->next;
 			depth--;
+			subtree_visited = true;
 		} else {
 			m = m->next;
+			subtree_visited = false;
 		}
 #endif
 	} while (m && m != root);
```

Comment:
The `m != root` check is no longer needed, looks like it's addressing this bug, but only for the root node.